cmd/client/client.go

```diff
@@ -0,0 +1,204 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Note: This is a quick-n-dirty client to enable me to test the server.
+//       This is _not_ intended to be the real client implementation.
+//       I'll implement a more robust client in a future change.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/adalton/teleport-exercise/certs"
+	"github.com/adalton/teleport-exercise/service/jobmanager/jobmanagerv1"
+
+	"google.golang.org/grpc"
+)
+
+const (
+	// On my local machine using the loopback interface, I measured the average
+	// time for things like start, stop, and query to take about 190ms.
+	// To allow for some transient delays and to support longer round-trips,
+	// we can be conservative at ~4x that value (800ms).  If that number is too
+	// conservative, we can collect more data and make a better estimate in the
+	// production enviornment in which we expect clients to use the service.
+	shortOperationTimeout = 800 * time.Millisecond
+
+	// With 1000 jobs, list takes ~200ms (~10ms more than the "short")
+	//      2000 jobs, list takes ~210ms (~10ms more than 1000 jobs)
+	//      3000 jobs, list takes ~220ms (~10ms more than 1000 jobs)
+	// Each job adds about 0.01 ms
+	// Assume a maximum of 100k jobs, that'd be ~1 second
+	listOperationTimeout = shortOperationTimeout + (1 * time.Second)
+)
+
+func timeOperation(operationName string) func() {
+	start := time.Now()
+
+	return func() {
+		fmt.Println(operationName, time.Since(start))
+	}
+}
+
+func msgContext(timeout time.Duration) (context.Context, func()) {
+	return context.WithTimeout(context.Background(), timeout)
+}
+
+func startContext() (context.Context, func()) {
+	return msgContext(shortOperationTimeout)
+}
+
+func stopContext() (context.Context, func()) {
+	return msgContext(shortOperationTimeout)
+}
+
+func queryContext() (context.Context, func()) {
+	return msgContext(shortOperationTimeout)
+}
+
+func listContext() (context.Context, func()) {
+	return msgContext(listOperationTimeout)
+}
+
+func streamContext() (context.Context, func()) {
+	// There's currently no upper limit on how long the streaming operation
+	// There's no need for the returned func, but keeping signature consistent
+	// enables us to change our minds in the future with a simpler code change.
+	return context.Background(), func() {}
+}
+
+func main() {
+	user := "client1"
+
+	if v, exists := os.LookupEnv("GRPC_USER"); exists {
+		user = v
+	}
+
+	fmt.Println("User:", user)
+
+	clientCert, clientKey, err := certs.ClientFactory(user)
+	if err != nil {
+		panic(err)
+	}
+
+	tc, err := certs.NewClientTransportCredentials(
+		certs.CACert,
+		clientCert,
+		clientKey,
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	conn, err := grpc.Dial("localhost:24482", grpc.WithTransportCredentials(tc))
```

Comment:
It seems that the server allows for the address to be configured via a flag now but `localhost:24482` is still hardcoded here

Reply:
Yep, there's a comment at the top of this file:

```go
// Note: This is a quick-n-dirty client to enable me to test the server.
//       This is _not_ intended to be the real client implementation.
//       I'll implement a more robust client in a future change.
```

As a result, I didn't take time to refactor this throw-away client.